content/blog/2019-08-tracing.md

```diff
@@ -0,0 +1,259 @@
+such as the client's IP address, the requested path, request method, headers,
+and so on. If we revisit the example above, with the edition of spans, we might
+see something like this:
+
+```plain
+DEBUG server{client.addr=106.42.126.8:56975}: accepted connection
+DEBUG server{client.addr=82.5.70.2:53121}: closing connection
+DEBUG server{client.addr=89.56.1.12:55601} request{path="/posts/tracing" method=GET}: received request
+DEBUG server{client.addr=111.103.8.9:49123}: accepted connection
+DEBUG server{client.addr=106.42.126.8:56975} request{path="/" method=PUT}: received request
+DEBUG server{client.addr=113.12.37.105:51342}: accepted connection
+ WARN server{client.addr=106.42.126.8:56975} request{path="/" method=PUT}: invalid request headers
+TRACE server{client.addr=106.42.126.8:56975} request{path="/" method=PUT}: closing connection
+```
+
+Notice how the events are annotated with spans that record the client IP
+address, and the request's path and method. Although multiple events are
+happening concurrently in different contexts, we can now follow the flow of the
+request from 106.42.126.8 through the system, and determine that it was the
+request containing the invalid headers that generated the warning.
+
+This machine-readable structured data also gives us the ability to consume diagnostic
+data in more sophisticated ways than simply formatting it to be read by a human.
+For example, we might also consume the above data by counting the number of
+requests recieved for different paths or HTTP methods. By looking at the
+structure of the span tree as well as at key-value data, we can even do things
+like recording the entire lifespan of a request only when it ended with an
+error.
+
+## Getting Started with Tracing
+
+`tracing` is available [on crates.io][tracing-crates]:
+```toml
+tracing = "0.1.5"
+```
+
+The easiest way to get started with `tracing` is to use the
+[`tracing::instrument`][inst] attribute on a function. This attribute will
+instrument the function to create and enter a new span when the function is
+called, with the function's arguments recorded as fields on that span. For
+example:
+
+```rust
+use tracing::instrument;
+
+#[instrument]
+pub async fn connect_to(remote: SocketAddr) -> io::Result<TcpStream> {
+    // ...
+}
+```
+
+`tracing` also provides a set of [function-like macros][macros] for constructing
+spans and events. Users of the `log` crate should note that `tracing`'s
+`trace!`, `debug!`, `info!`, `warn!` and `error!` macros are a superset of the
+similarly-named macros in `log` and should be drop-in compatible:
+
+```rust
+use log::info;
+
+info!("hello world!");
+```
+
+```rust
+use tracing::info;
+
+info!("hello world!");
+```
+
+The more idiomatic style, however, is to use these macros to record structured
+data rather than unstructured messages. For example:
+
+```rust
+use tracing::trace;
+
+let bytes_read = ...;
+let num_processed = ...;
+
+// ...
+
+trace!(bytes_read, messages = num_processed);
```

Comment:
worried the lack of a message field in this example could be confusing to readers who try to add a message or use structured data in a macro call that already has a message.